Test package Race detected Warning when DoHttps

I got race detected when trying sample code to try testing package
```go
type PluginConfig struct {
}

func (cfg PluginConfig) Access(pdk *pdk.PDK) {
	h, _ := pdk.Request.GetHeader("X-Hi")
	pdk.Response.SetHeader("x-hello-from-go", h)
}
func TestPlugin(t *testing.T) {
	env, err := test.New(t, test.Request{
		Method: "GET",
		Url:    "http://example.com?q=search&x=9",
		Headers: map[string][]string{
			"X-Hi": {"Go says Hi!"},
		},
	})
	assert.NoError(t, err)

	env.DoHttps(&PluginConfig{})
	assert.Equal(t, 200, env.ClientRes.Status)
	assert.Equal(t, "Go says Hi!", env.ClientRes.Headers.Get("x-hello-from-go"))
}

```
the race error shown
```
Running tool: /opt/homebrew/opt/go/libexec/bin/go test -timeout 5s -run ^TestPlugin$ github.com/barockok/kong-warden -race -v

=== RUN   TestPlugin
    /Users/zidni.mubarok/Documents/Barock-Nitip/Code/simple-abac/docker/kong/warden/test.go:579: Access
==================
WARNING: DATA RACE
Write at 0x00c0001c0088 by goroutine 8:
  github.com/Kong/go-pdk/test.(*TestEnv).DoHttp()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/test/test.go:635 +0x104
  github.com/Kong/go-pdk/test.(*TestEnv).DoHttps()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/test/test.go:645 +0x48
  github.com/barockok/kong-warden.TestPlugin()
      /Users/zidni.mubarok/Documents/Barock-Nitip/Code/simple-abac/docker/kong/warden/warden_test.go:72 +0x194
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1446 +0x188
  testing.(*T).Run.func1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1493 +0x40

Previous read at 0x00c0001c0088 by goroutine 9:
  github.com/Kong/go-pdk/test.(*TestEnv).IsRunning()
      <autogenerated>:1 +0x40
  github.com/Kong/go-pdk/bridge/bridgetest.MockFunc.func1()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/bridge/bridgetest/bridgetest.go:141 +0xc4

Goroutine 8 (running) created at:
  testing.(*T).Run()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1493 +0x55c
  testing.runTests.func1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1846 +0x90
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1446 +0x188
  testing.runTests()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1844 +0x6c0
  testing.(*M).Run()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1726 +0x870
  main.main()
      _testmain.go:51 +0x2fc

Goroutine 9 (running) created at:
  github.com/Kong/go-pdk/bridge/bridgetest.MockFunc()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/bridge/bridgetest/bridgetest.go:118 +0xe8
  github.com/Kong/go-pdk/test.New()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/test/test.go:206 +0x2b4
  github.com/barockok/kong-warden.TestPlugin()
      /Users/zidni.mubarok/Documents/Barock-Nitip/Code/simple-abac/docker/kong/warden/warden_test.go:63 +0x154
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1446 +0x188
  testing.(*T).Run.func1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1493 +0x40
==================
==================
WARNING: DATA RACE
Write at 0x00c0001c00b8 by goroutine 8:
  github.com/Kong/go-pdk/test.(*Response).merge()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/test/test.go:156 +0x60
  github.com/Kong/go-pdk/test.(*TestEnv).DoResponse()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/test/test.go:593 +0xcc
  github.com/Kong/go-pdk/test.(*TestEnv).DoHttp()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/test/test.go:636 +0x158
  github.com/Kong/go-pdk/test.(*TestEnv).DoHttps()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/test/test.go:645 +0x48
  github.com/barockok/kong-warden.TestPlugin()
      /Users/zidni.mubarok/Documents/Barock-Nitip/Code/simple-abac/docker/kong/warden/warden_test.go:72 +0x194
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1446 +0x188
  testing.(*T).Run.func1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1493 +0x40

Previous read at 0x00c0001c00b8 by goroutine 9:
  github.com/Kong/go-pdk/test.(*TestEnv).IsRunning()
      <autogenerated>:1 +0x40
  github.com/Kong/go-pdk/bridge/bridgetest.MockFunc.func1()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/bridge/bridgetest/bridgetest.go:141 +0xc4

Goroutine 8 (running) created at:
  testing.(*T).Run()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1493 +0x55c
  testing.runTests.func1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1846 +0x90
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1446 +0x188
  testing.runTests()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1844 +0x6c0
  testing.(*M).Run()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1726 +0x870
  main.main()
      _testmain.go:51 +0x2fc

Goroutine 9 (running) created at:
  github.com/Kong/go-pdk/bridge/bridgetest.MockFunc()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/bridge/bridgetest/bridgetest.go:118 +0xe8
  github.com/Kong/go-pdk/test.New()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/test/test.go:206 +0x2b4
  github.com/barockok/kong-warden.TestPlugin()
      /Users/zidni.mubarok/Documents/Barock-Nitip/Code/simple-abac/docker/kong/warden/warden_test.go:63 +0x154
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1446 +0x188
  testing.(*T).Run.func1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1493 +0x40
==================
==================
WARNING: DATA RACE
Write at 0x00c0001c00c0 by goroutine 8:
  github.com/Kong/go-pdk/test.(*Response).merge()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/test/test.go:159 +0x90
  github.com/Kong/go-pdk/test.(*TestEnv).DoResponse()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/test/test.go:593 +0xcc
  github.com/Kong/go-pdk/test.(*TestEnv).DoHttp()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/test/test.go:636 +0x158
  github.com/Kong/go-pdk/test.(*TestEnv).DoHttps()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/test/test.go:645 +0x48
  github.com/barockok/kong-warden.TestPlugin()
      /Users/zidni.mubarok/Documents/Barock-Nitip/Code/simple-abac/docker/kong/warden/warden_test.go:72 +0x194
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1446 +0x188
  testing.(*T).Run.func1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1493 +0x40

Previous read at 0x00c0001c00c0 by goroutine 9:
  github.com/Kong/go-pdk/test.(*TestEnv).IsRunning()
      <autogenerated>:1 +0x40
  github.com/Kong/go-pdk/bridge/bridgetest.MockFunc.func1()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/bridge/bridgetest/bridgetest.go:141 +0xc4

Goroutine 8 (running) created at:
  testing.(*T).Run()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1493 +0x55c
  testing.runTests.func1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1846 +0x90
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1446 +0x188
  testing.runTests()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1844 +0x6c0
  testing.(*M).Run()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1726 +0x870
  main.main()
      _testmain.go:51 +0x2fc

Goroutine 9 (running) created at:
  github.com/Kong/go-pdk/bridge/bridgetest.MockFunc()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/bridge/bridgetest/bridgetest.go:118 +0xe8
  github.com/Kong/go-pdk/test.New()
      /Users/zidni.mubarok/.go/pkg/mod/github.com/!kong/go-pdk@v0.8.0/test/test.go:206 +0x2b4
  github.com/barockok/kong-warden.TestPlugin()
      /Users/zidni.mubarok/Documents/Barock-Nitip/Code/simple-abac/docker/kong/warden/warden_test.go:63 +0x154
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1446 +0x188
  testing.(*T).Run.func1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1493 +0x40
==================
    /Users/zidni.mubarok/Documents/Barock-Nitip/Code/simple-abac/docker/kong/warden/testing.go:1319: race detected during execution of test
--- FAIL: TestPlugin (0.00s)
=== CONT
    /Users/zidni.mubarok/Documents/Barock-Nitip/Code/simple-abac/docker/kong/warden/testing.go:1319: race detected during execution of test
FAIL
FAIL	github.com/barockok/kong-warden	0.223s

```

## system info :

go-pdk : version  v0.8.0
go : version go1.20.7

--- a/bridge/bridgetest/bridgetest.go
+++ b/bridge/bridgetest/bridgetest.go
@@ -110,10 +110,14 @@ type mockEnvironment interface {
 	Handle(method string, args_d []byte) []byte
 	Errorf(format string, args ...interface{})
 	IsRunning() bool
+	SubscribeStatusChange(ch chan<- string)
 }
 
 func MockFunc(e mockEnvironment) net.Conn {
 	conA, conB := net.Pipe()
+
+	statusCh := make(chan string)
+	e.SubscribeStatusChange(statusCh)
 
 	go func() {
 		for {
@@ -138,8 +142,12 @@ func MockFunc(e mockEnvironment) net.Conn {
 				break
 			}
 
-			if !e.IsRunning() {
-				break
+			select {
+			case msg := <-statusCh:
+				if msg == "finnished" {
+					return
+				}
+			default: // do nothing
 			}
 		}
 	}()

--- a/test/test.go
+++ b/test/test.go
@@ -178,13 +178,14 @@ const (
 )
 
 type TestEnv struct {
-	t          *testing.T
-	state      envState
-	pdk        *pdk.PDK
-	ClientReq  Request
-	ServiceReq Request
-	ServiceRes Response
-	ClientRes  Response
+	t           *testing.T
+	state       envState
+	stateChange chan<- string
+	pdk         *pdk.PDK
+	ClientReq   Request
+	ServiceReq  Request
+	ServiceRes  Response
+	ClientRes   Response
 }
 
 // New creates a new test environment.
@@ -234,6 +235,17 @@ func (e TestEnv) Errorf(format string, args ...interface{}) {
 
 func (e TestEnv) IsRunning() bool {
 	return e.state == running
+}
+
+func (e *TestEnv) Finish() {
+	e.state = finished
+	if e.stateChange != nil {
+		e.stateChange <- "finished"
+	}
+}
+
+func (e *TestEnv) SubscribeStatusChange(ch chan<- string) {
+	e.stateChange = ch
 }
 
 // Internal use.  Handles a PDK request from the plugin under test.
@@ -426,7 +438,7 @@ func (e *TestEnv) Handle(method string, args_d []byte) []byte {
 		e.noErr(proto.Unmarshal(args_d, &args))
 		e.ClientRes.Status = int(args.Status)
 		if args.Headers != nil {
-			e.state = finished
+			e.Finish()
 			e.ClientRes.Body = args.Body
 			headers := bridge.UnwrapHeaders(args.Headers)
 			mergeHeaders(e.ClientRes.Headers, headers)

